### DISCUSSION TITLE
Setting up environment for CI/CD

### DISCUSSION THREAD

**Discussion by Pandawan** [hoodie/notify-rust#142] [Q&A]
(Continuing from #140)

I've setup a simple repository with docker to try and get a minimal reproducible example:

Dockerfile
```Dockerfile
FROM rust:1.59

WORKDIR /usr/src/example
COPY . .

RUN cargo install --path .

RUN cargo build

RUN apt-get update && apt-get install -y libdbus-1-dev

CMD ["cargo", "run"]
```

main.rs
```rust
use notify_rust::Notification;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    Notification::new()
        .summary("Title")
        .body("Body")
        .show()?;
    Ok(())
}
```

When running this docker container (with the appropriate cargo setup and `notify-rust  = "4.5.8"`), I get the error:
> thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { kind: Zbus(Io(Os { code: 2, kind: NotFound, message: "No such file or directory" })) }', src/main.rs:4:55

I'm assuming this means that my environment is missing some kind of config file (or maybe something else, given that unix uses files for everything). It seems like your [ci.yml](https://github.com/hoodie/notify-rust/blob/main/.github/workflows/ci.yml) file (for the `linux_zbus` job) doesn't have anything specific so I'm unsure what I could be missing.

Any insight would be appreciated!

**Comment by hoodie** [comment]
seems we're continuing in #140 after all 😁

**Reply by Pandawan** [reply]
Seems like the discussion in #140 has drifted off more into zbus vs dbus-rs?
I'm more looking for a general answer as to how to even get this working haha

**Reply by hoodie** [reply]
Maybe redundant question but, have you tried this with `RUST_BACKTRACE=full` yet? I'm afraid the error message is not indicative of the actual problem.

**Reply by Pandawan** [reply]
Yeah, unfortunately I don't get much more than that... 
It may have to do with there being no `DBUS_SESSION_BUS_ADDRESS` as was suggested in #140? (On my docker example, this environment variable is not set). 
But I don't see you setting that in the ci.yml file so it's probably something else.

**Reply by zeenix** [reply]
> It may have to do with there being no `DBUS_SESSION_BUS_ADDRESS` as was suggested in #140?

Yes and you need the bus running as well. If you want your app to be talking to your desktop session, you need to provide access to the session bus socket on your host to the docker container and set the env.

As for a CI example, the [zbus CI](https://gitlab.freedesktop.org/dbus/zbus/-/blob/main/.gitlab-ci.yml#L118) could give you some clues.

**Reply by Pandawan** [reply]
Gotcha, I'll look into those, thanks! It's just strange that notify-rust's ci doesn't have any of these and yet is able to test it... Unless I'm missing something.

**Reply by zeenix** [reply]
@Pandawan best of luck. @hoodie can say if I'm wrong but I'm guessing the notify's tests are not e2e and hence not talking to the service on the bus.

**Reply by hoodie** [reply]
oh yes, my bad, I forgot that, sorry!
That's what the original server was for. I usually do manual tests locally on.
That is partially also necessary because KDE/Gnome/Xfce etc have slightly different behavior when displaying the notifications. so the actual sending is not tested in CI.

**Reply by Pandawan** [reply]
Gotcha, so I guess what @zeenix said with zbus' CI is the way to go. Thanks!

**Comment by hoodie** [comment]
Thing is, something needs to provide that org.freedesktop.notification interface, otherwise you'll just get errors upon sending too.
